### PR TITLE
Give the manifest env() helper an injectable env seam (#484)

### LINKED ISSUE #484: Inject an Env seam into the manifest env() Jinja function
## Context

The manifest `env()` Jinja function reads the process environment directly at `src/manifest/mod.rs:85`:

```rust
fn env_var(name: &str) -> std::result::Result<String, Error> {
    match std::env::var(name) {
```

The function has three distinct outcomes — present, `NotPresent`, and `NotUnicode` — and the last two are precisely the cases that are awkward to reach without mutating the real environment. Under the [AGENTS.md testing mandate](https://github.com/leynos/netsuke/blob/main/AGENTS.md), in-process mutation is no longer available, so these branches currently have no legitimate route to coverage.

The doc comment on `env_var` already advertises a `test_support::env::VarGuard` that mutates global state; that example must go with the implementation.

## Required work

- Give `env_var` an injected environment, and carry it on the value registered into the Jinja environment so `manifest::from_str` can supply one.
- Supply the process-backed default reader (`process_env_reader`) at the manifest public boundary.
- Rewrite the `env_var` doc example to construct a `MockEnv` rather than a mutating guard.
- Migrate `tests/manifest_env_tests.rs` onto the injected seam, covering all three outcomes including the non-UTF-8 branch.

## Acceptance criteria

- No `std::env::var` call remains in `src/manifest/mod.rs`.
- `tests/manifest_env_tests.rs` no longer acquires `EnvLock` or mutates the process environment.
- The `NotUnicode` branch is covered by a test.
- `make check-fmt`, `make lint`, and `make test` pass.

---

**Amended:** the seam is a closure (`EnvReader`), not `&impl mockable::Env`.
The requirement above has been rewritten to match; see the discussion below for
the reasoning. `mockable` is a dev-dependency, and the proportionality argument
is that a production dependency is a large price for reading one variable when
the repository already uses closure seams of the same shape in `stdlib::path`,
`output_mode`, `output_prefs`, and `theme`.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

`env_var` — backing the `env()` Jinja helper — read the process environment directly:

```rust
fn env_var(name: &str) -> std::result::Result<String, Error> {
    match std::env::var(name) { ... }
}
```

Its three outcomes could only be reached by mutating global state, which the [AGENTS.md testing mandate](https://github.com/leynos/netsuke/blob/main/AGENTS.md) no longer permits. One was unreachable in practice regardless: fabricating a non-UTF-8 value in the live environment needs platform-specific `OsString` surgery, so the `NotUnicode` branch and its distinct Jinja error kind had **no coverage at all**.

The function's doc comment also advertised a `test_support::env::VarGuard` example that mutates global state; that goes with it.

## Approach

Split resolution into `env_var_with`, taking a `read_env` closure — consistent with the existing seam in `runner::process::ninja_program` and the two added in #486 and #487. A closure rather than a trait object: keyed lookup, one caller, and the mandate permits a narrow closure where a trait object would be disproportionate.

## Coverage

Six cases, none mutating anything:

- a present variable resolves to its value
- **an empty value is returned rather than treated as missing** — an empty value is a value, not an absence
- both failures map to their documented Jinja error kind
- **the two failure kinds stay distinct** — a missing variable is a template authoring error, whereas a non-UTF-8 value is an environment problem the author cannot fix in the template. Collapsing them onto one kind would misdirect whoever reads the failure, and nothing previously stopped that
- the requested name reaches the seam unaltered and appears in the message

## Checked against upstream RFCs

[ortho-config RFC 0001](https://github.com/leynos/ortho-config/blob/main/docs/rfcs/0001-field-level-environment-aliases.md) governs field-level environment aliases for *configuration structs*. It has no bearing on the manifest `env()` helper, which is Netsuke's own template surface, so the closure seam here is not in tension with the `EnvSource` abstraction being added upstream.

## Verification

All gates pass: `check-fmt`, `lint`, `typecheck`, `test` (1194 nextest), `markdownlint`, `nixie`. CodeScene delta: no issues.

Closes #484.
Refs #496.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Introduce an injectable environment lookup seam for the manifest env() helper to enable full testing of all resolution outcomes without mutating process state.

Enhancements:
- Add env_var_with helper that accepts a closure for environment variable resolution while preserving existing env_var behavior and error mapping semantics.

Tests:
- Add dedicated env_function tests that cover successful resolution, empty values, distinct failure kinds, and propagation of variable names through error messages without touching the real environment.

### CODE REVIEW

File: src/manifest/mod.rs [hunk 1]
```diff
@@ -70,19 +70,29 @@ pub enum ManifestLoadStage {
...
-/// // guard restores prior value on drop
+/// let value = env_var_with("FOO", |_| Ok(String::from("bar")));
+/// assert_eq!(value.expect("FOO"), "bar");
 /// ```
-fn env_var(name: &str) -> std::result::Result<String, Error> {
-    match std::env::var(name) {
+fn env_var_with<F>(name: &str, read_env: F) -> std::result::Result<String, Error>
+where
+    F: FnOnce(&str) -> std::result::Result<String, std::env::VarError>,
+{
```
[the remark this answers is not in the document]
Comment (pandalump) [reply src/manifest/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please reply confirming this. Otherwise, please provide a follow-up recommendation.

[the remark this answers is not in the document]
Comment (pandalump) [reply src/manifest/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please reply confirming this. Otherwise, please provide a follow-up recommendation.

File: src/manifest/mod.rs [hunk 2]
```diff
@@ -70,19 +70,29 @@ pub enum ManifestLoadStage {
 /// Returns the variable's value or a structured error that mirrors Jinja's
 /// failure modes, ensuring templates halt when a variable is missing or not
 /// valid UTF-8.
+fn env_var(name: &str) -> std::result::Result<String, Error> {
+    env_var_with(name, |key| std::env::var(key))
```
[the remark this answers is not in the document]
Comment (pandalump) [reply src/manifest/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please reply confirming this. Otherwise, please provide a follow-up recommendation.

**Comment by buzzybee-df12** [comment]
@coderabbitai review

File: src/manifest/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (leynos) [reply src/manifest/mod.rs]: The `mockable` half of that acceptance criterion is withdrawn, and the issue text was wrong to state it.

`mockable` sits in `[dev-dependencies]` and reaches the crate only through `test_support`. `manifest` is production code with a public API, so using `mockable::Env` there requires moving `mockable` into `[dependencies]` — and it brings `mockall` with it:

```
mockable v3.0.0
├── mockall v0.11.4
└── tracing v0.1.44
```

Shipping a mocking framework in release builds of a build tool is not a reasonable price for replacing a two-line closure alias. The same question came up on #502 and was resolved the same way, with the reasoning recorded there.

[AGENTS.md](https://github.com/leynos/netsuke/blob/main/AGENTS.md) permits this explicitly: *"A narrow closure seam of the same shape is acceptable where a trait object would be disproportionate."* `EnvReader` is `Fn(&str) -> Result<String, VarError>`, which is `mockable::Env::raw`'s shape exactly.

What the criterion was really after — that the registration path be testable without mutating the process — is met, as your codegraph analysis on the sibling threads confirms.

**Comment by buzzybee-df12** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### docs/developers-guide.md (1)

`837-839`: _📐 Maintainability & Code Quality_ | _🟡 Minor_ | _⚡ Quick win_
**Restrict the ownership claim to production code.**
`tests/manifest_env_tests.rs` constructs `EnvReader` directly with `Arc::new`. Therefore, “Nothing else constructs one except the public entry points” is too broad. State that production callers use the public entry points, while tests may construct injected readers directly.
Triage: `[type:docstyle]`

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@docs/developers-guide.md` around lines 837 - 839, Update the ownership
statement near manifest::from_str_named to scope it to production callers: say
production code uses the public entry points, while tests such as
manifest_env_tests.rs may construct injected EnvReader instances directly with
Arc::new.

**Comment by pandalump** [comment]
@coderabbitai review

File: docs/developers-guide.md [hunk 3]
```diff
@@ -1174,6 +1180,36 @@ pattern documented in the
...
+#### Ownership and permitted call sites
+
+- The reader is owned by `manifest::from_str_named`, which is the only place
+  the `env()` function is registered. Nothing else constructs one except the
+  public entry points.
+- `process_env_reader()` is the sole production supplier and the only place
+  `std::env::var` appears in the module.
+- Tests use `from_str_with_env`. They must not reach past it to the leaf
+  mapper: covering `env_var_with` alone would leave the registration untested,
+  which is the gap that made the earlier process-mutating tests necessary.
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/developers-guide.md]: Fixed in `1aa9c9b`. Both claims were false, and I checked rather than assuming:

- `tests/manifest_env_tests.rs:24,67` builds readers with `Arc::new` — which is the point of the seam, so the guide was describing a restriction that does not and should not exist.
- `src/manifest/tests/env_function.rs` calls `env_var_with` directly throughout.

The guide now describes the split that actually exists, with the reason each layer is needed: integration tests exercise *registration* — that the reader reaches the `env()` function Jinja calls — while unit tests cover *error mapping* at the leaf, where the non-UTF-8 branch is reachable without platform-specific `OsString` surgery.

Also renamed that section's `Composition rules` heading to `Reader composition rules`; it collided with the one at line 503 under MD024, which `make markdownlint` caught.

File: src/manifest/env_reader.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+/// # Examples
+///
+/// ```rust,no_run
+/// use netsuke::manifest::process_env_reader;
+///
+/// // Reads whatever the process has. Not executed: the result depends on the
+/// // caller's environment, so running it would make the doctest a hostage to
+/// // whatever CI happens to export.
+/// let reader = process_env_reader();
+/// let _ = reader("PATH");
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/manifest/env_reader.rs]: Fixed in `1aa9c9b`.

Worth noting the distinction, since I pushed back on the same suggestion elsewhere today: here `reader(\"PATH\")` returns `Result<String, VarError>`, which **is** `#[must_use]`, so `drop(...)` is right. On #502 the same recommendation was made for a bare `Option<String>`, and `Option` is not `#[must_use]` — only `Result` is. The rule that holds in both cases: `Result` in statement position wants `drop`; `Option` does not.

**Comment by buzzybee-df12** [comment]
@coderabbitai review

File: src/manifest/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pandalump) [reply src/manifest/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please reply confirming this. Otherwise, please provide a follow-up recommendation.

[the remark this answers is not in the document]
Comment (pandalump) [reply src/manifest/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please reply confirming this. Otherwise, please provide a follow-up recommendation.

File: src/manifest/env_reader.rs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+/// // caller's environment, so running it would make the doctest a hostage to
+/// // whatever CI happens to export.
+/// let reader = process_env_reader();
+/// drop(reader("PATH"));
+/// ```
+#[must_use]
+#[expect(
+    clippy::disallowed_methods,
+    reason = "composition root: supplies the process environment to the env() Jinja helper"
+)]
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/manifest/env_reader.rs:40]: Respectfully declining — this contradicts the policy that #505 established and which is now on `main`.

`docs/developers-guide.md` states three dispositions, the first being:

> **Composition roots** in `src/` keep a permanent site-level expectation naming the seam they supply. These are the sanctioned ambient boundary.

`process_env_reader` is exactly that: the one place the manifest's `env()` helper is bound to the live process, with `env_var_with` taking the reader as a parameter everywhere else. There is no work item because there is no migration pending — the seam already exists and this function is its production end.

On `expect` versus `allow`, the choice is deliberate and documented: `expect` goes *unfulfilled* — and warns — the moment a site stops tripping the lint, so a migrated file fails the gate until its annotation is removed. `allow` would go stale invisibly. That property is why the annotation is `expect` even though the exception is permanent, and it has already earned its keep: rebasing these branches over #505 produced exactly those unfulfilled warnings where code had moved, and each one pointed at an annotation that needed relocating.

Leaving as-is.

File: tests/makefile_test_target.rs [hunk 6]
```diff
@@ -710,11 +710,23 @@ fn behavioural_nextest_config_binds_the_env_binaries_to_the_serial_env_group() -
...
+        .split("binary(")
+        .skip(1)
+        .filter_map(|rest| rest.split(')').next())
+        .collect();
+    members.sort_unstable();
+    ensure!(
+        members == ["env_path_tests", "ninja_env_tests"],
+        "serial-env should cover exactly the two PATH- and NINJA_ENV-mutating \
+         binaries; found {members:?} in filter {filter:?}"
+    );
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/makefile_test_target.rs:730]: Correct, and now tracked as #524 rather than fixed here.

The file is 732 lines. It was **720 on `main`** before this branch touched it; #501 adds 18 lines — an exact-membership assertion for the `serial-env` nextest group, replacing a weaker presence check. Splitting a 732-line file is grossly disproportionate to an 18-line change, and would bury this PR's subject (an injected environment seam for the manifest `env()` helper) under an unrelated refactor.

Worth noting why it drifted this far unnoticed: Whitaker enforces module length for the crate's own modules but not for crate-root integration test files, so `make lint` passes at 732 lines. I have recorded that in the issue as well — the tooling gap is arguably the more useful half of the finding, since it is what let a 400-line rule be exceeded by 80%.

**Comment by buzzybee-df12** [comment]
@coderabbitai review

File: docs/developers-guide.md [hunk 7]
```diff
@@ -538,12 +538,12 @@ governs the non-doctest pass only, and deliberately stays small:
...
-the threaded in-process harness. The `EnvLock`, `EnvVarGuard`, and `CwdGuard`
-utilities described in [Test isolation utilities](#test-isolation-utilities),
-and the `#[serial]` markers on the tests in the three binaries above, remain
-necessary because the coverage workflow still drives an in-process runner.
+nextest runs each test in its own process, so environment and working-directory
+mutations cannot leak between tests the way they can under the threaded
+in-process harness. The `EnvLock`, `EnvVarGuard`, and `CwdGuard` utilities
+described in [Test isolation utilities](#test-isolation-utilities), and the
+`#[serial]` markers on the tests in the three binaries above, remain necessary
+because the coverage workflow still drives an in-process runner.
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/developers-guide.md:1159]: Confirmed resolved — you marked this addressed in `6c1e3e1`, and the latest round produced no new findings on it.

File: src/manifest/env_reader.rs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+///
+/// # Examples
+///
+/// ```rust
+/// use netsuke::manifest::process_env_reader;
+///
+/// let reader = process_env_reader();
+/// // Reads whatever the process has; absent variables report `NotPresent`.
+/// assert!(reader("NETSUKE_DEFINITELY_UNSET_VARIABLE").is_err());
+/// ```
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/manifest/env_reader.rs]: Confirmed resolved in `6c1e3e1` — the doctest is `rust,no_run`, so it no longer depends on what the caller's environment happens to export.

File: src/manifest/env_reader.rs [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+/// Returns the variable's value or a structured error that mirrors Jinja's
+/// failure modes, ensuring templates halt when a variable is missing or not
+/// valid UTF-8.
+/// Resolve `name` through `read_env`, mapping failures to Jinja errors.
+///
+/// Separated from [`env_var`] so all three outcomes — present, absent, and
+/// non-UTF-8 — can be exercised without mutating the process environment. The
+/// non-UTF-8 branch is otherwise unreachable from a test: fabricating such a
+/// value in the live environment needs platform-specific `OsString` surgery,
+/// and the AGENTS.md testing mandate forbids in-process mutation regardless.
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/manifest/env_reader.rs]: Confirmed resolved in `6c1e3e1` — the stale intra-doc link to the removed `env_var` is gone.

**Comment by leynos** [comment]
Re the two pre-merge warnings on this PR.

**Linked Issues check** — taken via the second option it offers, "update #484 to approve the EnvReader abstraction". #484's *body* still specified `mockable::DefaultEnv`, which is what the check reads; the amendment had only been recorded as a comment beneath it. The body now states the closure seam as the requirement, so the issue and the implementation agree.

Worth recording accurately, because my earlier reasoning on this was incomplete: I had argued that `mockable` cannot be a production dependency because it pulls `mockall` into release builds. That is only true if the `mock` feature travels with it — `mockall` is `optional = true` behind `mock = ["dep:mockall"]`, and this crate is edition 2024, so a `[dependencies]` entry without the feature would ship nothing extra. The decision to keep a closure therefore rests on proportionality rather than on a hard blocker: a production dependency is a large price for reading one variable, when `stdlib::path`, `output_mode`, `output_prefs`, and `theme` already use closure seams of the same shape. That is a judgement, and #525 records the equivalent open question for path types.

**User-Facing Documentation** — accepted and not yet done. `from_str_with_env`, `EnvReader`, and `process_env_reader` are public API and `docs/users-guide.md` says nothing about them. I am out of context to write it properly in this session rather than declining it; it needs a short programmatic-API section covering when to pass a reader and what `process_env_reader` supplies.

**Comment by leynos** [comment]
The accepted users-guide item is now done: docs/users-guide.md gains an "Inject the environment reader for tests and embedding" subsection beside the existing `env(name)` paragraph, naming `from_str`, `from_str_with_env`, and `process_env_reader`, restating the missing/non-Unicode failure mapping, mirroring the executable doctest, and cross-referencing the developers' guide's "Manifest `env()` reader" section.

**Comment by pandalump** [comment]
@coderabbitai review

File: docs/users-guide.md [hunk 10]
```diff
@@ -432,6 +432,45 @@ Both helpers accept:
...
+);
+let manifest = from_str_with_env(yaml, &reader).expect("parse");
+assert!(format!("{:?}", manifest.targets[0].name).contains("release"));
+```
+
+This snippet mirrors the executable doctest on `from_str_with_env` in the API
+documentation, rather than the YAML-only examples elsewhere in this guide.
+Contributors extending this seam should also read the developers' guide's
+[Manifest `env()` reader](developers-guide.md#manifest-env-reader) section,
+which covers the ownership boundary and composition rules.
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md]: Removed in the latest commit. The contributor-facing sentence is gone from the users' guide — the section now ends with the doctest-mirror note and stays embedder-facing. The same commit also fixes the two documents the Developer Documentation warning named: the developers' guide's ownership bullet (the caller owns the reader; `from_str_named` borrows it and `Arc::clone`s it into the single registration site) and netsuke-design.md §4.4 (now describes the `EnvReader` boundary rather than direct process reads).

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### tests/makefile_test_target.rs (1)

`169-498`: _📐 Maintainability & Code Quality_ | _🟠 Major_ | _🏗️ Heavy lift_
**Split the Rustflags contract tests into a sibling module.**
Move the Rustflags helpers and tests into a dedicated module. Keep both Rust files

below 400 lines. Add a `//!` module comment to the new file.
As per coding guidelines, “No single Rust source file may exceed 400 lines.”

As per path instructions, “Files must not exceed 400 lines.”

##### Proposed module boundary

```diff
+#[path = "makefile_test_target/rustflags_contract.rs"]
+mod rustflags_contract;
+
-const RUSTFLAGS_PREFIX: &str = "RUSTFLAGS=\"";
-// Move the RustflagsCase helpers and Rustflags contract tests to the new module.
```

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@tests/makefile_test_target.rs` around lines 169 - 498, Move the
Rustflags-specific constants, RustflagsCase, helper functions, and related tests
from the current test module into a sibling Rust module, preserving their
behavior and test coverage. Add a module-level //! documentation comment to the
new file, and ensure both the original file and the new module remain below 400
lines.

_Sources: Coding guidelines, Path instructions, Learnings_

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat o11y, code safety, documentation and validation coverage as in scope).

### Assessment against linked issues

| Issue | Objective | Addressed | Explanation |
| --- | --- | --- | --- |
| [#484](https://github.com/leynos/netsuke/issues/484) | Introduce an injectable environment seam for the manifest env() helper, including removing direct std::env::var usage from src/manifest/mod.rs and supplying a mockable DefaultEnv at the manifest public boundary. | ❌ | The PR introduces env\_var\_with(name, read\_env) and has env\_var delegate to it via a closure, which does provide an injectable seam. However, env\_var still calls std::env::var inside src/manifest/mod.rs via env\_var\_with(name, |

**Comment by leynos** [comment]
On the pre-merge table's Developer Documentation warning: that row predates dce4728, which already updated netsuke-design.md §4.4 (it now describes `env()` reading through the injected `EnvReader`, names `process_env_reader()` and `from_str_with_env()`, and states the caller-owns-the-reader model) and corrected the developers' guide's `from_str_named` ownership claim in the same commit. The refreshed table on the next round should reflect it.

Separately, the makefile_test_target decomposition requested against this PR now lives in #526, stacked on this branch, closing #524.

**Comment by pandalump** [comment]
@coderabbitai review
